### DISCUSSION TITLE
MuseSampler API

### DISCUSSION THREAD

**Discussion by isolin** [musescore/MuseScore#17801] [Ideas]
### Your idea

I aosolutely love Muse Hub Sounds. I see a strong parallel to Blender 3D which has Cycles as its awesome off-line renderer. Cycles is fully integrated into Blender, but it can also work in standalone mode. I would like to ask for the same with the MuseScore sequencer in connection with MuseHub Sounds. Would it be possible to expose its functionality over an API? Just as an array of timed events to get rid of the layouting and formatting that is part of the written score, but not really necessary for the playback. It could even work as an option in the MuseScore command line, but instad of taking `mscz` it would take a simple file format... or as a library one could call.

 If anyhing like that already exists, please let me know.

### Problem to be solved

Using MuseScore's playback and export functionality as a plain Sampler without the necessity to design and format a digital score. Direct control of articulations, effect values, etc.

### Additional context

My current workaround would be to wrap the desired input into the `mscz` format (all the files inside the zip archive), mainly dealing with the `mscx` xml-based format, but I consider that a huge overhead.

**Comment by cbjeukendrup** [comment]
MuseSampler can indeed be loaded by other programs. There is no official documentation (yet?), but probably it's much more valuable anyway to take MuseScore's source code as an example. You can basically write your own app that does whatever you want with Muse Sampler. Or you can integrate it into some DAW... 
(just talking about _technical_ possibilities here; note that since MuseScore's source code is GPL3, derivative works will need to be licensed similarly afaik.)

I don't expect that MuseScore will support yet another file format for "playback only", since music notation _is_ simply MuseScore's reason to exist. However, a separate app dedicated to MuseSampler might be a nice idea, if someone would like to build that.

**Comment by isolin** [comment]
Thank you very much for the answer. Yes a separate app under an open license looks like a good idea to me as well. I scrolled through the MuseScore code to find references to the sampler, but so far I failed. Could you please give me some pointers where to find them?

**Comment by cbjeukendrup** [comment]
The most directly interesting stuff is in `src/framework/musesampler/internal/libhandler.h`, which is basically about how the library is _loaded_. You could basically copy this file in its entirety.
If you open the project in Qt Creator, you can see in which files `libhandler.h` is included, namely `musesamplersequencer.h`, `musesamplerresolver.h` and `musesamplerwrapper.h`. In those files you see how the library is _used_. 
For how playback events are 'scraped' from the score, see the files in `src/engraving/playback`.

**Comment by isolin** [comment]
Thank you, those look like good starting points!

**Comment by bkunda** [comment]
This is worth discussing (thank you @isolin), but it doesn't really fit here as an issue for MuseScore's development.
So I shall convert this to a discussion. 
Thank you!

**Reply by lyrra** [reply]
I think this should be an issue. Community project?
Such an network API could enable separation of sequencer and score-editor (as we're currently doing in Mu3).
This in turn would enable a rock-solid score-editor, because then MuseSampler or VSTs can crash and burn as much as it wants without bringing the whole system down.
Although if Mu4 was to do this, then I guess there wont be much difference betwen score-editor and staffpad.

**Comment by swesterfeld** [comment]
I wonder if this could be made available as LV2 or CLAP plugin (or if it must be, VST) so that one could use the sampler from any DAW?

**Comment by charlesneimog** [comment]
@cbjeukendrup Sorry to bother you with this.

But I had a working version for PureData from version `0.5` of MuseSampler, but it does not work with version `0.6`. 

I can't init the library, 

**Question**: There was some change in the init process?

I am just changing how `LOGE` and `LOGW` works and also replace  `MuseSamplerLibHandler(const io::path_t& path)` to `MuseSamplerLibHandler(std::string path)`. I am using just `libhandler.h` and `apitypes.h`. My init is basically this

``` cpp

class t_MuseSampler {
  public:
    t_object xObj;

    mu::musesampler::MuseSamplerLibHandlerPtr MuseSounds = nullptr;
/// etc

// ===================

void *myinit(){
    t_MuseSampler *x = (t_MuseSampler *)pd_new(MuseSampler);
    x->MuseSounds = std::make_shared<MuseSamplerLibHandler>(path);
    if (!x->MuseSounds->isValid()) {
        pd_error(NULL, "MuseSounds not found, if you are sure you have it please report");
        return nullptr;
    }
    bool ok = x->MuseSounds->init();

    if (!ok) { // error here
        return nullptr;
    }

// etc ....

``` 

https://github.com/charlesneimog/pd-musesampler

Sorry again, any info will be of great help! :)

**Reply by cbjeukendrup** [reply]
That looks like it should work (because it's exactly similar to MuseScore's MuseSamplerResolver::doInit). We'd probably need a bit more information about what exactly doesn't work. Could you elaborate about that? And at which step does it fail exactly?

**Reply by charlesneimog** [reply]
In the function `init`, this is `true` for me:

``` cpp
if (initLib() != ms_Result_OK) {
    LOGE() << "Could not init lib";
    return false;
}
``` 

I am waiting to try this in another machine, maybe something with my machine (but MuseScore works normally).

**Reply by cbjeukendrup** [reply]
Okay, I found the problem. Unfortunately, it turns out to be outside your control. 

MuseSampler checks whether the name of the executable that is loading it is `mscore`.

At least that is the case on macOS. When I right-click my `MuseScore 4.app`, click "Show contents", navigate to `Content/MacOS`, and rename `mscore` to something else, and then double-click it, it fails to init MuseSampler. When I rename it back to `mscore`, MuseSampler works again.

**Reply by charlesneimog** [reply]
Bad news, :(.

Thanks for the info and help!
